In [1]:
import numpy as np

print("--- RUČNO RAČUNANJE KVANTNIH STANJA KROZ LINEARNU ALGEBRU ---")

# 1. DEFINIŠEMO KLASIČNA STANJA KAO VEKTORE (Baza Z)
# Stanje |0> je vertikalni vektor [1, 0]
stanje_0 = np.array([[1], [0]])

# 2. DEFINIŠEMO MATRICU HADAMARDOVE KAPIJE (H)
# Formula za H kapiju je 1/sqrt(2) * [[1, 1], [1, -1]]
H_kapija = (1 / np.sqrt(2)) * np.array([[1, 1], 
                                        [1, -1]])

print("\nMatrica Hadamardove kapije (H):\n", np.round(H_kapija, 3))

# 3. RUČNO MNOŽENJE MATRICE I VEKTORA (Kreiranje superpozicije)
# Množimo H matricu sa vektorom stanja |0> da dobijemo stanje |+>
stanje_plus = np.dot(H_kapija, stanje_0)

print("\nKonačni vektor stanja |+> nakon ručnog množenja:\n", np.round(stanje_plus, 3))

# 4. DOKAZ VEROVATNOĆE (Hezenbergov princip)
# Verovatnoća da je kubit 0 ili 1 se dobija kvadratom amplituda stanja
verovatnoca_nule = stanje_plus[0][0]**2
verovatnoca_jedinice = stanje_plus[1][0]**2

print(f"\nIzračunata verovatnoća stanja |0>: {verovatnoca_nule * 100:.1f}%")
print(f"Izračunata verovatnoća stanja |1>: {verovatnoca_jedinice * 100:.1f}%")


--- RUČNO RAČUNANJE KVANTNIH STANJA KROZ LINEARNU ALGEBRU ---

Matrica Hadamardove kapije (H):
 [[ 0.707  0.707]
 [ 0.707 -0.707]]

Konačni vektor stanja |+> nakon ručnog množenja:
 [[0.707]
 [0.707]]

Izračunata verovatnoća stanja |0>: 50.0%
Izračunata verovatnoća stanja |1>: 50.0%


Šta ovaj kod matematički dokazuje? 

Kod dokazuje proces formiranja kvantne superpozicije na čistom algebarskom nivou, bez korišćenja eksternih kvantnih simulatora, kroz direktno množenje vektora početnog stanja sa unitarnom matricom kapije.

Kako je dobijena superpozicija u vektorskom obliku? Množenjem matrice Hadamardove kapije sa vektorom kolone klasičnog stanja |0> [[0], [1]], dobili smo novi vektor stanja |+> sa amplitudama 
[[0.707], [0.707]].

Šta nam ovi brojevi govore o verovatnoći sistema? 

Kvadratom ovih amplituda (0.707²) dobijamo tačno vrednosti od 0.5, što je direktan matematički dokaz da se sistem nalazi u stabilnoj superpoziciji sa simetričnom verovatnoćom od tačno 50% za oba bazična stanja.